lib/fd.c

```diff
@@ -0,0 +1,45 @@
+// SPDX-FileCopyrightText: 2024, Skyler Ferrante <sjf5462@rit.edu>
+// SPDX-License-Identifier: BSD-3-Clause
+
+/**
+ * To protect against file descriptor omission attacks, we open the std file
+ * descriptors with /dev/null if they are not already open. Code is based on
+ * fix_fds from sudo.c.
+ */
+
+#include "prototypes.h"
+
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+static void check_fd (int fd);
+
+void
+check_fds (void)
+{
+	/**
+	 * Make sure stdin, stdout, stderr are open
+	 * If they are closed, set them to /dev/null
+	 */
+	check_fd (STDIN_FILENO);
+	check_fd (STDOUT_FILENO);
+	check_fd (STDERR_FILENO);
+}
+
+static void
+check_fd (int fd)
+{
+	int  devnull;
+
+	if (fcntl (fd, F_GETFL, 0) != -1)
+		return;
+
+	devnull = open ("/dev/null", O_RDWR, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH);
+	if (devnull == -1)
+		abort ();
+
+	if (dup2 (devnull, fd) == -1)
+		abort ();
```

Comment:
Sorry, for not seeing this before.

I've been thinking about what you said, and yeah dup2 is redundant, and I think we can make it simpler, keeping it readable:


```c
static void
check_fd(int fd)
{
	if (fcntl (fd, F_GETFL, 0) != -1)
		return;

	if (open("/dev/null", O_RDWR, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH) != fd)
		abort ();
}
```

Does it look good to you?


BTW, you don't need to add those spaces before opening parenthesis.  We're slowly removing them, especially in new code, to eventually get rid of them without churning.

Reply:
Also, I think octal is more readable for the mode.

I would write it as
```c
open("/dev/null", O_RDWR, 0644)
```

But, according to open(2), the mode is only used when creating a file.

```
            The  mode  argument specifies the file mode bits to be ap‐
            plied when a new file is created.  If neither O_CREAT  nor
            O_TMPFILE is specified in flags, then mode is ignored (and
            can  thus be specified as 0, or simply omitted).  The mode
            argument must be supplied if O_CREAT or O_TMPFILE is spec‐
            ified in flags; if it  is  not  supplied,  some  arbitrary
            bytes from the stack will be applied as the file mode.
```

Thus, the mode is ignored, and can be specified as 0, or omitted.

```c
open("/dev/null", O_RDWR)
```

Reply:
>  dup2 is redundant, and I think we can make it simpler, keeping it readable

Hmm, I'm much more hesitant about that. It is redundant if the OS gives us new file descriptors starting at our lowest free slot. This is true on Linux, and I assume all Unix Operating systems. However, since dup2 is guaranteed to be safe if newfd and oldfd is equal, it's probably best to keep it. What do you think?

> BTW, you don't need to add those spaces before opening parenthesis

Okay, I can remove them.

Reply:
> > dup2 is redundant, and I think we can make it simpler, keeping it readable
> 
> Hmm, I'm much more hesitant about that. It is redundant if the OS gives us new file descriptors starting at our lowest free slot. This is true on Linux, and I assume all Unix Operating systems.

It's mandated by POSIX:

<https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_14>

```
All functions that open one or more file descriptors shall, unless
specified otherwise, atomically allocate the lowest numbered available
(that is, not already open in the calling process) file descriptor at
the time of each allocation. Where a single function allocates two file
descriptors (for example, pipe() or socketpair()), the allocations may
be independent and therefore applications should not expect them to have
adjacent values or depend on which has the higher value.
```

> However, since dup2 is guaranteed to be safe if newfd and oldfd is equal, it's probably best to keep it. What do you think?

It's guaranteed to be redundant, so I don't see why we should keep it.

> 
> > BTW, you don't need to add those spaces before opening parenthesis
> 
> Okay, I can remove them.

Reply:
Okay, I wasn't sure if it was mandated anywhere. And I will take your suggestion for the open as well.